app/controllers/api/v1/base_controller.rb

```diff
@@ -0,0 +1,89 @@
+# frozen_string_literal: true
+
+module Api
+  module V1
+    class BaseController < ActionController::API
+      include CanCan::ControllerAdditions
+      include RequestTimeouts
+      include Pagy::Backend
+      include JsonApiPagination
+
+      RESULTS_PER_PAGE = 50
+
+      check_authorization
+
+      attr_accessor :current_api_user
+
+      before_action :authenticate_user
+
+      rescue_from Exception, with: :error_during_processing
+      rescue_from CanCan::AccessDenied, with: :unauthorized
+      rescue_from ActiveRecord::RecordNotFound, with: :not_found
+
+      private
+
+      def authenticate_user
+        return if (@current_api_user = request.env['warden'].user)
+
+        if api_key.blank?
+          # An anonymous user
+          @current_api_user = Spree::User.new
+          return
+        end
+
+        return if (@current_api_user = Spree::User.find_by(spree_api_key: api_key.to_s))
+
+        invalid_api_key
+      end
+
+      def current_ability
+        Spree::Ability.new(current_api_user)
+      end
+
+      def api_key
+        request.headers["X-Api-Token"] || params[:token]
+      end
+
+      def error_during_processing(exception)
+        Bugsnag.notify(exception)
+
+        render status: :unprocessable_entity,
+               json: json_api_error(exception.message, backtrace: exception.backtrace)
+      end
+
+      def invalid_resource!(resource = nil)
+        render status: :unprocessable_entity,
+               json: json_api_invalid(I18n.t(:invalid_resource, scope: "spree.api"), resource&.errors)
+      end
+
+      def invalid_api_key
+        render status: :unauthorized,
+               json: json_api_error(I18n.t(:invalid_api_key, key: api_key, scope: "spree.api"))
+      end
+
+      def unauthorized
+        render status: :unauthorized,
+               json: json_api_error(I18n.t(:unauthorized, scope: "spree.api"))
+      end
+
+      def not_found
+        render status: :not_found,
+               json: json_api_error(I18n.t(:resource_not_found, scope: "spree.api"))
+      end
+
+      def json_api_error(message, **options)
+        error_response = { errors: [{ detail: message }] }
+        if options[:backtrace] && (Rails.env.development? || Rails.env.test?)
```

Comment:
is this something we actually use or is useful?

Reply:
I threw this in while I was doing a little bit of tidying up (whilst mostly just copying this code across from the old BaseController). 

It's really useful both in development and when running unit tests (while doing TDD for example). Normally you just get an error message with no context, and then you have to spend a bunch of extra time digging for more information. If you get it straight away it actually makes it a lot faster to fix things. It doesn't interfere with the content in the top-level `errors` key.

```
{
  errors: <error message>,
  meta: {
    backtrace: <backtrace lines> # Only in dev/test
  }
}
```

Reply:
ah, nice! I remember having that problem.